scripts/test_env.py

```diff
@@ -0,0 +1,191 @@
+        os.environ["SUMERAGI_TRUSTED_PEERS"] = f"[{','.join(self.trusted_peers)}]"
+
+    def run(self):
+        self.peers[0].run(is_genesis=True)
+        for peer in self.peers[1:]:
+            peer.run()
+
+class Peer:
+    """
+    A single bootstrapped peer. Could be a genesis node or a regular peer.
+    Should not be run directly, but rather as a part of a Network.
+    """
+    def __init__(self, setup_dir: pathlib.Path, host_addr: ipaddress.IPv4Address, nth: int):
+        self.nth = nth
+        self.name = f"iroha{nth}"
+        self.p2p_port = 1337 + nth
+        self.api_port = 8080 + nth
+        self.telemetry_port = 8180 + nth
+        self.tokio_console_port = 5555 + nth
+        self.setup_dir = setup_dir
+        self.host_addr = host_addr
+        self.generate_key_pair()
+        logging.info("Peer %s generated key pair", self.name)
+        
+    def generate_key_pair(self):
+        kagami = subprocess.run([f"{self.setup_dir}/kagami","crypto", "-j"], capture_output=True)
+        if kagami.returncode:
+            logging.error("Kagami failed to generate a key pair.")
+            sys.exit(1)
+        str_keypair = kagami.stdout
+        json_keypair = json.loads(str_keypair)
+        # public key is a string, private key is a json object
+        self.public_key = json_keypair['public_key']
+        self.private_key = json.dumps(json_keypair['private_key'])
+
+    def run(self, is_genesis: bool = False):
+        logging.info("Running peer %s...", self.name)
+        peer_dir = self.setup_dir.joinpath(f"peers/{self.name}")
+        os.mkdir(peer_dir)
+        os.mkdir(peer_dir.joinpath("storage"))
+
+        # Copying the binary to have unique names for each peer
+        iroha_bin = f"{peer_dir}/{self.name}"
+        shutil.copy2(f"{self.setup_dir}/peers/iroha", iroha_bin)
+
+        os.environ["KURA_BLOCK_STORE_PATH"] = str(peer_dir.joinpath("storage"))
+        os.environ["LOG_FILE_PATH"] = str(peer_dir.joinpath("log.json"))
+        os.environ["MAX_LOG_LEVEL"] = "TRACE"
+        os.environ["IROHA_PUBLIC_KEY"] = self.public_key
+        os.environ["IROHA_PRIVATE_KEY"] = self.private_key
+        os.environ["SUMERAGI_DEBUG_FORCE_SOFT_FORK"] = "false"
+        os.environ["TORII_P2P_ADDR"] = f"{self.host_addr}:{self.p2p_port}"
+        os.environ["TORII_API_URL"] = f"{self.host_addr}:{self.api_port}"
+        os.environ["TORII_TELEMETRY_URL"] = f"{self.host_addr}:{self.telemetry_port}"
+        os.environ["TOKIO_CONSOLE_ADDR"] = f"{self.host_addr}:{self.tokio_console_port}"
+
+        genesis_arg = "--submit-genesis" if is_genesis else ""
+        # FD never gets closed
+        log_file = open(peer_dir.joinpath("out.log"), "w")
+        # These processes are created detached from the parent process already
+        subprocess.Popen([iroha_bin, f"{genesis_arg}"],
+                    stdout=log_file, stderr=subprocess.STDOUT)
```

Comment:
Notice that in the `exec` in my script, I rename the processes to be `iroha0` and so on. Consider adding this here.

Reply:
But that's what actually is being done here. I tried to do child renaming with `exec`, but `subprocess` was being weird about it and many hours were lost. So I opted for the dumber approach where I rename and copy the binaries into each peer's directory. The only difference is that the process name includes full relative path to it when viewed with `ps` or `top`.

Reply:
But that's what actually is being done here. I tried to do child renaming with `exec`, but `subprocess` was being weird about it and many hours were lost. So I opted for the dumber approach where I rename and copy the binaries into each peer's directory. The only difference is that the process name includes full relative path to it when viewed with `ps` or `top`.